python/ray/data/_internal/arrow_serialization.py

```diff
@@ -63,3 +101,199 @@ def _register_arrow_json_parseoptions_serializer():
         ),
         deserializer=lambda args: pajson.ParseOptions(*args),
     )
+
+
+# Register custom Arrow data serializer to work around zero-copy slice pickling bug.
+# See https://issues.apache.org/jira/browse/ARROW-10739.
+def _register_arrow_data_serializer():
+    """Custom reducer for Arrow data that works around a zero-copy slicing pickling
+    bug by using the Arrow IPC format for the underlying serialization.
+
+    Background:
+        Arrow has both array-level slicing and buffer-level slicing; both are zero-copy,
+        but the former has a serialization bug where the entire buffer is serialized
+        instead of just the slice, while the latter's serialization works as expected
+        and only serializes the slice of the buffer. I.e., array-level slicing doesn't
+        propagate the slice down to the buffer when serializing the array.
+
+        All that these copy methods do is, at serialization time, take the array-level
+        slicing and translate them to buffer-level slicing, so only the buffer slice is
+        sent over the wire instead of the entire buffer.
+
+    See https://issues.apache.org/jira/browse/ARROW-10739.
+    """
+    import ray
+    import pyarrow as pa
+
+    if os.environ.get(RAY_DISABLE_CUSTOM_ARROW_DATA_SERIALIZATION) == "1":
+        import logging
+
+        logger = logging.getLogger(__name__)
+        logger.info(
+            "Disabling custom Arrow data serialization. This may result in bloated "
+            "serialization of Arrow tables!"
+        )
+        return
+
+    context = ray._private.worker.global_worker.get_serialization_context()
+    # Register custom reducer for Arrow Arrays.
+    array_types = _get_arrow_array_types()
+    for array_type in array_types:
+        context._register_cloudpickle_reducer(array_type, _arrow_array_reduce)
+    # Register custom reducer for Arrow ChunkedArrays.
+    context._register_cloudpickle_reducer(pa.ChunkedArray, _arrow_chunkedarray_reduce)
+    # Register custom reducer for Arrow RecordBatches.
+    context._register_cloudpickle_reducer(pa.RecordBatch, _arrow_recordbatch_reduce)
+    # Register custom reducer for Arrow Tables.
+    context._register_cloudpickle_reducer(pa.Table, _arrow_table_reduce)
+
+
+def _get_arrow_array_types() -> List[type]:
+    """Get all Arrow array types that we want to register a custom serializer for."""
+    try:
+        import pyarrow as pa
```

Comment:
In which case do we expect this? The pyarrow is a required install when installing Ray.

Reply:
I moved this check to happen once, but we need this check when registering these serializers because `pyarrow` is not a dependency when installing the minimal Ray package via `pip install ray`, it's part of the `ray[data]` extra requirements that would be installed with `pip install ray[data]`: https://github.com/ray-project/ray/blob/3c4302906fd0d3958ccf2588ad37dd5b9912b1d9/python/setup.py#L226

So we need this serializer to short-circuit when `pyarrow` is not installed, since this serializer registration code will happen on `import ray`.